From 88d65608a2bea19d803da47b0b75b7d3a3b86531 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Caol=C3=A1n=20McNamara?= <caolanm@redhat.com>
Date: Thu, 21 Feb 2019 15:57:22 +0000
Subject: [PATCH] menu of currency combobox in format-cells is too narrow

we want the combobox to be narrower than it wants to be.  To make that happen
we have only the option of shrinking the cell renderer of the combobox area.
And that is also used by the menu.

Setting a small value of e.g. 1 works to let the combobox not request more
width than that, and the combobox will correctly render within the wider size
it actually gets. But then the menu is a min width menu, which is undesirable,
we want the menu to be as wide as it can be.

So calculate what part of the combobox belongs to the cell area and set
the widest cell area we can within the overall combobox width, resulting
in the widest menu we can get.

Change-Id: Ie3e9960a320a70471ac21d2a88f32632cafa951f
---
 vcl/unx/gtk3/gtk3gtkinst.cxx | 28 +++++++++++++++++++++++++---
 1 file changed, 25 insertions(+), 3 deletions(-)

diff --git a/vcl/unx/gtk3/gtk3gtkinst.cxx b/vcl/unx/gtk3/gtk3gtkinst.cxx
index 808c1c6..b0e1573 100644
--- a/vcl/unx/gtk3/gtk3gtkinst.cxx
+++ b/vcl/unx/gtk3/gtk3gtkinst.cxx
@@ -6732,9 +6732,31 @@ public:
         // tweak the cell render to get a narrower size to stick
         GList* cells = gtk_cell_layout_get_cells(GTK_CELL_LAYOUT(m_pComboBox));
         GtkCellRenderer* cell = static_cast<GtkCellRenderer*>(cells->data);
-        GtkRequisition size;
-        gtk_cell_renderer_get_preferred_size(cell, m_pWidget, &size, nullptr);
-        gtk_cell_renderer_set_fixed_size(cell, nWidth, size.height);
+
+        if (nWidth != -1)
+        {
+            // this bit isn't great, I really want to be able to ellipse the text in the comboboxtext itself and let
+            // the popup menu render them in full, in the interim ellipse both of them
+            g_object_set(G_OBJECT(m_pTextRenderer), "ellipsize", PANGO_ELLIPSIZE_MIDDLE, nullptr);
+
+            // to find out how much of the width of the combobox belongs to the cell, set
+            // the cell and widget to the min cell width and see what the difference is
+            int min;
+            gtk_cell_renderer_get_preferred_width(cell, m_pWidget, &min, nullptr);
+            gtk_cell_renderer_set_fixed_size(cell, min, -1);
+            gtk_widget_set_size_request(m_pWidget, min, -1);
+            int nNonCellWidth = get_preferred_size().Width() - min;
+
+            // now set the cell to the max width which it can be within the
+            // requested widget width
+            gtk_cell_renderer_set_fixed_size(cell, nWidth - nNonCellWidth, -1);
+        }
+        else
+        {
+            g_object_set(G_OBJECT(m_pTextRenderer), "ellipsize", PANGO_ELLIPSIZE_NONE, nullptr);
+            gtk_cell_renderer_set_fixed_size(cell, -1, -1);
+        }
+
         g_list_free(cells);
 
         gtk_widget_set_size_request(m_pWidget, nWidth, nHeight);
-- 
2.20.1

